### PR TITLE
feat(dspark): DeepSeek-V4 Flash/Pro DSpark self-speculative drafter

### PR DESCRIPTION
**Pull request by popfido**
## DeepSeek-V4 Flash/Pro DSpark self-speculative decoding

Adds [DSpark](https://github.com/deepseek-ai/DeepSpec) self-speculative decoding for **DeepSeek-V4 (Flash and Pro)**. Self-contained against `main` — it brings its own DSpark round loop + dispatch wiring (the generic foundation also shared with the separate Gemma 4 DSpark PR #1465, but this PR does not depend on it). One config drives both sizes. Output is byte-identical to base greedy decoding.

> Independent of #1465: both PRs introduce the generic DSpark foundation (`speculative/dspark.py` round loop, the `dspark` dispatch kind, the CLI flag). They can be reviewed and merged in either order; whichever lands second has a trivial overlap on those shared files.

### How it works
DSpark is an EAGLE-style drafter that conditions on the target's intermediate hidden states, drafts a block of tokens biased by a low-rank Markov head, and exposes a confidence head; the target verifies losslessly. The DeepSeek-V4 draft is the windowed-MLA + hash-MoE + Hyper-Connection realization that ships under the `mtp.*` namespace of the base fp8/fp4 checkpoint.

### Target change
`mlx_vlm/models/deepseek_v4/language.py`: add `capture_layer_ids` to `DeepseekV4Model`/`LanguageModel` so the target exposes per-layer hiddens at the DSpark target layers. Each captured hidden is the Hyper-Connection copies reduced via `h.mean(axis=2)` — matching the reference `Transformer.forward`. Capture is **additive**: base logits and the existing MTP last-hidden path are unchanged.

### Drafter — reuse over re-implementation
`mlx_vlm/speculative/drafters/deepseek_v4_dspark/`. Reuse-vs-port decided by numerical micro-probes, not assumption:

| Component | Decision | Parity vs reference |
|-----------|----------|---------------------|
| `DeepseekV4MoE` (hash/score MoE) | **reuse** | exact (0.0) |
| `HyperConnection` / `HyperHead` / `hc_expand` | **reuse** | ~1e-6 |
| Markov / confidence heads | **reuse** (bias-free for V4) | identical |
| Windowed MLA cross-attention + RoPE/RMSNorm | **port** (no in-tree equivalent) | — |

The drafter exposes both the reference windowed API (`forward_spec`/`advance`) and the mlx-vlm round-loop contract (`draft_block`/`reset`/`make_cache`). It also corrects a shared-expert SwiGLU clamp the official spec applies but `DeepseekV4MoE` currently omits.

### Wiring + real-weight path
- Routes `model_type=deepseek_v4_dspark` to the package and maps it to the `dspark` round-loop kind (same model_type→package convention as `deepseek_v4_mtp`). CLI: `--draft-kind dspark`, `--draft-confidence-threshold`.
- `split.py` extracts the `mtp.*` draft + shared `embed`/`head` from the base fp8/fp4 checkpoint and **dequantizes to bf16** (the reference runs a bf16/fp32 drafter; the ported attention does manual grouped `wo_a` math), mirroring `deepseek_v4_mtp.split`.

```sh
python -m mlx_vlm.speculative.drafters.deepseek_v4_dspark.split \
  --model deepseek-ai/DeepSeek-V4-Flash-DSpark --output ./dsv4-flash-dspark-drafter
mlx_vlm.generate --model deepseek-ai/DeepSeek-V4-Flash \
  --draft-model ./dsv4-flash-dspark-drafter --max-tokens 256 --temperature 0 \
  --prompt "Explain speculative decoding in 3 sentences."
```

### Validation
The validator is **numerical parity against [dspark-mlx](https://github.com/popfido/dspark-mlx)** — the validated MLX port of the official `inference/model.py` (which is CUDA/tilelang-bound and can't run on Mac). On identical random weights, the ported `forward_spec` matches the reference drafter:

- tokens **identical**; logits **~1.8e-7**; confidence **~2.4e-7**
- shared-expert clamp tripped: tokens identical, confidence ~3.4e-4
- `advance`/multi-step windowing: logits ~1.5e-7

`mlx_vlm/tests/test_deepseek_v4_dspark.py`: config mapping (Flash/Pro), sanitize key roundtrip, `draft_block` shapes, single/batch **losslessness** (output == base greedy via a Markov-stub target), routing, the split path, and oracle-gated `forward_spec`/`advance` parity (skipped if `dspark_mlx` isn't importable).

### Test plan
- [x] `pytest mlx_vlm/tests/test_deepseek_v4_dspark.py` — 16 passed, 1 skipped (CI); 11 passed with the parity oracle on path
- [x] `pytest mlx_vlm/tests/test_speculative.py` — no regressions (166 passed / 1 skipped together with the DSpark suite + oracle)
- [x] pre-commit (black 24.2.0 / isort / autoflake) clean
- [ ] **Real-weight run on a DeepSeek-V4-DSpark checkpoint** — gated behind `DSPARK_DEEPSEEK_V4_SOURCE` (`test_real_checkpoint_splits_loads_and_drafts`). The fp8/fp4 dequant in `split` follows the base model's quantization-config inverse but is validated on-machine, not in CI. **Benchmark numbers (speedup / acceptance) to follow once run on real weights.**

### CODE REVIEW

**Comment by Lazarus-931** [comment]
closing, since existent inside `speculative/drafters/deepseek_v4_dspark/`
